packages/sources/tp-icap/src/endpoint/price.ts

```diff
@@ -0,0 +1,128 @@
+import {
+  PriceEndpoint,
+  priceEndpointInputParameters,
+} from '@chainlink/external-adapter-framework/adapter'
+import {
+  WebSocketTransport,
+  WebSocketRawData,
+} from '@chainlink/external-adapter-framework/transports/websocket'
+import { makeLogger } from '@chainlink/external-adapter-framework/util'
+
+import Decimal from 'decimal.js'
+
+import { TPICAPWebsocketGenerics } from '../types'
+
+const logger = makeLogger('TPICAPPrice')
+
+const isNum = (i: any) => typeof i === 'number'
+
+let providerDataStreamEstablishedUnixMs: number
+
+export const transport: WebSocketTransport<TPICAPWebsocketGenerics> =
+  new WebSocketTransport<TPICAPWebsocketGenerics>({
+    url: ({ adapterConfig: { WS_API_ENDPOINT } }) => WS_API_ENDPOINT,
+    handlers: {
+      open: (connection, { adapterConfig: { WS_API_USERNAME, WS_API_PASSWORD } }) => {
+        logger.debug('Opening WS connection')
+
+        return new Promise((resolve, reject) => {
+          connection.on('message', (data: WebSocketRawData) => {
+            const { msg, sta } = JSON.parse(data.toString())
+            if (msg === 'auth' && sta === 1) {
+              logger.info('Got logged in response, connection is ready')
+              providerDataStreamEstablishedUnixMs = Date.now()
+              resolve()
+            } else {
+              reject(new Error('Failed to make WS connection'))
+            }
+          })
+          const options = {
+            msg: 'auth',
+            user: WS_API_USERNAME,
+            pass: WS_API_PASSWORD,
+            mode: 'broadcast',
+          }
+          connection.send(JSON.stringify(options))
+        })
+      },
+      message: (message) => {
+        logger.debug({ msg: 'Received message from WS', message })
+
+        const providerDataReceivedUnixMs = Date.now()
+
+        if (!('msg' in message) || message.msg === 'auth') return []
+
+        const { fvs, rec, sta } = message
+
+        if (!fvs || !rec || sta !== 1) {
+          logger.error({ msg: 'Missing expected field `fvs` or `rec` from `sub` message', message })
+          return []
+        }
+
+        const { ACTIV_DATE, ASK, BID, MID_PRICE, TIMACT } = fvs
+
+        const base = rec.slice(5, 8)
```

Comment:
it seems we should be using the `CCY1` and `CCY2` fields in the `fvs` object instead of parsing `rec` (which is the event name)

Reply:
Using `rec` moving forward with a `requestTransform` to construct the code from base and quote if the requested `base` is 3 characters long. Specific source requests should override `base` with the full code.